Sources/Navigation/Extensions/WKFrameInfoExtension.swift

```diff
@@ -0,0 +1,110 @@
+//
+//  WKNavigationActionExtension.swift
+//
+//  Copyright © 2023 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import WebKit
+import os.log
+
+public extension WKFrameInfo {
+
+    internal static var defaultMainFrameHandle = "4"
+
+    // prevent exception if private API keys go missing
+    override func value(forUndefinedKey key: String) -> Any? {
+        assertionFailure("valueForUndefinedKey: \(key)")
+        return nil
+    }
+
+    @nonobjc var handle: String {
+#if DEBUG
+        String(describing: (self.value(forKey: "handle") as? NSObject)!.value(forKey: "frameID")!)
+#else
+        self.isMainFrame ? Self.defaultMainFrameHandle : "iframe"
+#endif
+    }
+
+    /// Safe Optional `request: URLRequest` getter:
+    /// .request of a new Frame can be `null`, see https://app.asana.com/0/0/1203965979591356/f
+    var safeRequest: URLRequest? {
+        _=WKFrameInfo.addSafetyCheckForSafeRequestUsageOnce
+        return self.perform(#selector(getter: request))?.takeUnretainedValue() as? URLRequest
+    }
+
+#if DEBUG
+    private static var ignoredRequestUsageSymbols = Set<String>()
```

Comment:
Definitely not blocking this PR, but given we use it in 3 places, we may want to abstract this mechanism for future use at some point of time (QWD?).

Reply:
maybe in future, given that it‘s not business code duplication I‘ll live with it for now